cpp/include/cudf/scalar/scalar_device_view.cuh

```diff
@@ -45,21 +46,12 @@ class scalar_device_view_base {
     return *_is_valid;
   }
 
-  /**
```

Comment:
So this is my attempt at making scalar_device_view a read-only object.  I made things const and eliminated write methods.

Reply:
What I'm worried about now is how to enable mutable scalars. They seem to share a lot of methods with the immutable ones. 

One way could be to use something similar to column_view i.e. an unexposed base class that has common methods and two derived classes: one immutable, one mutable. A problem with that is the number of classes needed. We'd need 3 classes for each data type: 1 base, 1 mutable, 1 immutable.

Another option is to derive the mutable class from the immutable class and add all non-const methods to it. Problem with that would be that downcasting an immutable class to its mutable child would effectively make it mutable. 

@jrhemstad is there some wisdom you can share based on your work on `column_view`?

Reply:
> Another option is to derive the mutable class from the immutable class and add all non-const methods to it. 

This is what I would have done for `column_view` if not for the children. 

> Problem with that would be that downcasting an immutable class to its mutable child would effectively make it mutable.

I'm not following on this one. I imagine the structure would be:
```
class scalar_device_view{
   // const stuff
};

class mutable_scalar_view : public scalar_device_view {
   // non-const stuff
};
```

```c++
scalar_device_view * v = new scalar_device_view(...);

mutable_scalar_view * m = static_cast<mutable_scalar_view*>(v); // this is illegal
```

Reply:
> ```c++
> scalar_device_view * v = new scalar_device_view(...);
> 
> mutable_scalar_view * m = static_cast<mutable_scalar_view*>(v); // this is illegal
> ```

Doesn't look like it can stop me.
https://onlinegdb.com/SyyI0gohH

Reply:
Yeah, there's nothing the compiler can do to stop you, it's still illegal/undefined behavior.

Reply:
The compiler will stop you from accidentally doing things improperly... which is I think the important part.  It's still a C derivative, you can do all manner of evil improper things if you try hard enough.

Reply:
In that case, this PR wouldn’t preclude a mutable scalar view. So I’ll approve